credentials/tls/certprovider/provider.go

```diff
@@ -0,0 +1,107 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package certprovider defines APIs for certificate providers in gRPC.
+//
+// Experimental
+//
+// Notice: All APIs in this package are experimental and may be removed in a
+// later release.
+package certprovider
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+)
+
+var (
+	// errProviderClosed may be returned from a call to KeyMaterial() when the
+	// underlying provider instance is closed.
+	errProviderClosed = errors.New("provider instance is closed")
+
+	// m is a map from name to provider builder.
+	m = make(map[string]Builder)
+)
+
+// Register registers the provider builder, whose name as returned by its
+// Name() method will be used as the name registered with this builder.
+func Register(b Builder) {
+	m[b.Name()] = b
+}
+
+// getBuilder returns the provider builder registered with the given name.
+// If no builder is registered with the provided name, nil will be returned.
+func getBuilder(name string) Builder {
+	if b, ok := m[name]; ok {
+		return b
+	}
+	return nil
+}
+
+// Builder creates a Provider.
+type Builder interface {
+	// Build creates a new provider with the provided config.
+	Build(StableConfig) Provider
+
+	// ParseConfig converts config input in a format specific to individual
+	// implementations and returns an implementation of the StableConfig
+	// interface.
+	// Equivalent configurations should return StableConfig types whose
+	// Canonical() method returns the same output.
+	ParseConfig(interface{}) (StableConfig, error)
+
+	// Name returns the name of providers built by this builder.
+	Name() string
+}
+
+// StableConfig wraps the method to return a stable provider configuration.
+type StableConfig interface {
```

Comment:
What does "stable" mean?

Reply:
"stable" means something that does not change. Proto-serialization is not stable. Proto-to-json is not stable. I have added a comment on `ParseConfig` about what this means. Do you think it will be good to add something here as well?

Reply:
+1 yes this needs a comment.  "Stable" is not a universal term.  It could mean "doesn't change once created", not the stricter definition you are using here of "all equivalent configurations have an identical StableConfig".